SC_redesign/Crop_and_Soil/soil/soil_temp.py

```diff
@@ -0,0 +1,544 @@
+#         weather: an instance of the Weather class
+#         time: an instance of the Time class
+#
+#     Returns:
+#         int: radiate, the radiation term for the temperature of bare soil
+#     """
+#
+#     H_day = weather.radiation[time.year - 1][time.day - 1]
+#     albedo = calc_albedo(soil, crop_type)
+#
+#     return (H_day * (1 - albedo) - 14) / 20
+#
+#
+# def calc_albedo(soil, crop_type):
+#     """
+#     Description:
+#         Calculates the daily albedo as a function of soil type, plant cover,
+#         and snow cover.
+#         "pseudocode_soil" S.1.A.9/10
+#
+#     Args:
+#         soil: an instance of the Soil class
+#         crop: an instance of the Crop class
+#
+#     Returns:
+#         int: soil albedo
+#     """
+#
+#     CV = crop_type.bio_AG
+#
+#     # "pseudocode_soil" S.1.A.10
+#     cover = exp(-0.00005 * CV)
+#
+#     # "pseudocode_soil" S.1.A.9
+#     return 0.23 * (1 - cover) + soil.soil_albedo * cover
+#
+#
+# def calc_bcv(crop_type, time):
+#     """
+#     Description:
+#         Calculates the weighting factor for ground cover
+#         "pseudocode_soil" S.1.A.11/12
+#
+#     Args:
+#         crop: an instance of the Crop class
+#         time: an instance of the Time class
+#
+#     Returns:
+#         int: bcv, the bio-cover weighting factor for ground cover
+#     """
+#
+#     CV = crop_type.bio_AG
+#     exp_part = exp(7.563 - 0.0001297 * (-CV))
+#
+#     bcv = CV / (CV + exp_part)
+#
+#     snow = 0
+#     # TODO: arbitrary snow flag - GitHub Issue #164
+#     if time.day > 335 or time.day < 59:
+#         albedo_snow = 0.8
+#         snow = 10 * albedo_snow
+#
+#     bcv_snow = (snow / (snow + exp(6.055 - 0.3002 * snow)))
+#
+#     return max(bcv, bcv_snow)
+from typing import Optional
+from math import exp, log
+
+from SC_redesign.Crop_and_Soil.soil.soil_data import SoilData
+
+"""
+This module is based on the "Soil Temperature" section of SWAT (1:1.3.3)
+"""
+class SoilTemp:
+
+    def __init__(self, soil_data: Optional[SoilData] = None):
+        self.data = soil_data or SoilData()
+
+    def daily_soil_temperature_update(self, solar_radiation: float, avg_temp: float, min_temp: float, max_temp: float,
+                                      plant_cover: float, snow_cover: float, avg_annual_air_temp: float) -> None:
+        """this is the main routine that updates the soil temperature
+
+        Args:
+            solar_radiation: solar radiation reaching the ground on a given day (MJ per square meter per day)
+            avg_temp: average temperature of a given day (degrees C)
+            min_temp: minimum temperature of a given day (degrees C)
+            max_temp: maximum temperature of a given day (degrees C)
+            plant_cover: total aboveground plant biomass and residue on a given day (kg per hectare)
+            snow_cover: water content of the snow cover on a given day (mm)
+            avg_annual_air_temp: average annual air temperature (degrees C)
```

Comment:
from the perspective of the simulation:

```suggestion
            solar_radiation: solar radiation reaching the ground on the current day (MJ per square meter per day)
            avg_temp: average temperature of the current day (degrees C)
            min_temp: minimum temperature of the current day (degrees C)
            max_temp: maximum temperature of the current day (degrees C)
            plant_cover: total aboveground plant biomass and residue on the current day (kg per hectare)
            snow_cover: water content of the snow cover on the current day (mm)
            avg_annual_air_temp: average annual air temperature (degrees C)
```

Reply:
Related: is `avg_annual_air_temp` the average for the entire calendar year, the current calendar year up to the current day (I hope not), or for the last 365 days?

This variable worries me for when we want the model to forecast into the future (not for v1).

Reply:
I don't know. The most detail I have found about `avg_annual_air_temp` is "average annual air temperature is calculated from  the long-term monthly maximum and minimum temperatures reported in the weather generator input (.wgn) file" on page 42 of SWAT